src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp

```diff
@@ -91,20 +94,95 @@ namespace DB
         else
             status = builder.AppendValues(internal_data.data(), internal_data.size(), reinterpret_cast<const uint8_t *>(arrow_null_bytemap_raw_ptr));
         checkStatus(status, write_column->getName(), format_name);
+    }
 
-        status = builder.Finish(&arrow_array);
-        checkStatus(status, write_column->getName(), format_name);
+    static void fillArrowArrayWithArrayColumnData(
+        const String & column_name,
+        ColumnPtr & nested_column,
+        const std::shared_ptr<const IDataType> & column_type,
+        std::shared_ptr<arrow::Array> arrow_array,
+        const PaddedPODArray<UInt8> * null_bytemap,
+        arrow::ArrayBuilder * array_builder,
+        String format_name)
+    {
+        const auto * column_array = static_cast<const ColumnArray *>(nested_column.get());
+        const bool is_column_array_nullable = column_array->getData().isNullable();
+        const IColumn & array_nested_column =
+            is_column_array_nullable ? static_cast<const ColumnNullable &>(column_array->getData()).getNestedColumn() :
+            column_array->getData();
+        const String column_array_nested_type_name = array_nested_column.getFamilyName();
+
+        const auto * column_array_type = static_cast<const DataTypeArray *>(column_type.get());
+        const DataTypePtr & array_nested_type =
+            is_column_array_nullable ? static_cast<const DataTypeNullable *>(column_array_type->getNestedType().get())->getNestedType() :
+            column_array_type->getNestedType();
+
+        const auto * arrow_type_it = std::find_if(internal_type_to_arrow_type.begin(), internal_type_to_arrow_type.end(),
+                                                  [=](auto && elem) { return elem.first == column_array_nested_type_name; });
+        if (arrow_type_it != internal_type_to_arrow_type.end())
+        {
+            std::shared_ptr<arrow::DataType> list_type = arrow::list(arrow_type_it->second);
+
+            const auto & internal_column = assert_cast<const ColumnArray &>(*nested_column);
+
+            arrow::ListBuilder & builder = assert_cast<arrow::ListBuilder &>(*array_builder);
+            arrow::ArrayBuilder * value_builder = builder.value_builder();
+            arrow::Status components_status;
+
+            const auto & offsets = internal_column.getOffsets();
+            const IColumn & data = internal_column.getData();
```

Comment:
The problem is here. If nested type is nullable then data will be ColumnNullable, but we want nested column. Try this:
```c++
const IColumn & data = is_column_array_nullable ? static_cast<const ColumnNullable &>(internal_column.getData()).getNestedColumn() : internal_column.getData();

```